tools/deployment_tool/contract.js

```diff
@@ -0,0 +1,323 @@
+        }
+
+        if (typeof addressGeneratorOrAddress === 'string') {
+            this.address = addressGeneratorOrAddress;
+            return this.address;
+        }
+
+        const addressGenerator = addressGeneratorOrAddress;
+        if (!addressGenerator) {
+            throw new Error(`addressGenerator not provided when generating address for ${this.contractName}`);
+        }
+
+        const address = addressGenerator.generateAddress();
+        this.address = address;
+        return this.address;
+    }
+
+    /**
+     * Get the address of an instantiated contract.
+     *
+     * @returns {string} The address of the instantiated contract.
+     */
+    getAddress() {
+        this._ensureState(STATE_INSTANTIATED);
+        return this.address;
+    }
+
+    /**
+     * Returns the previously linked bytecode.
+     *
+     * @returns {string} The linked bytecode of the contract.
+     */
+    linkedBytecode() {
+        this._ensureStateOneOf([STATE_LINKED, STATE_INSTANTIATED]);
+        return this.contractBytecode;
+    }
+
+    /**
+     * Instantiate the contract by calculating the data of the contract creation transaction.
+     * This includes encoding the provided constructor arguments.
+     * Freezes the object to prevent any further changes.
+     *
+     * @return {string} The transaction data for contract creation.
+     */
+    instantiate() {
+        this._ensureState(STATE_NEW);
+
+        if (this.constructorABI && !this.constructorArgs) {
+            throw new Error(`Expected constructor arguments for constract ${this.contractName}`);
+        }
+        if (this.constructorArgs && !this.constructorABI) {
+            throw new Error(`Provided arguments for contract ${this.contractName}, but no constructorAbi is set.`);
+        }
+
+        this._linkBytecode();
+
+        let constructorData = this.linkedBytecode();
+        constructorData += this._encodeConstructorArguments();
+
+        this.constructorData = constructorData;
+        this._state = STATE_INSTANTIATED;
+        return constructorData;
+    }
+
+    /**
+     * Returns a reference to the contract, which can be used as a placeholder for
+     * a contract address in constructor arguments.
+     *
+     * @returns {object} A reference to this contract.
+     */
+    reference() {
+        return {
+            __type: TYPE_CONTRACT_REFERENCE,
+            contract: this,
+        };
+    }
+
+    /**
+     * Tries to interpret the provided constructor argument as a contract reference,
+     * and returns the referenced contract if it is one. See {@link Contract#reference}.
+     *
+     * @param {*} constructorArg A constructor argument that is possibly a contract reference.
+     *
+     * @returns {Contract|null} The referenced contract.
+     */
+    _getReferenceContract(constructorArg) {
+        if (!((typeof constructorArg === 'object') && (constructorArg !== null)) || constructorArg.__type !== TYPE_CONTRACT_REFERENCE) {
```

Comment:
Maybe something like the below. I am sure this can inspire you to do something readable 😄 
```js
if(!isReference(constructorArg)) {
    return null;
}


function isReference(constructorArg) {
    if (!isInstance(constructorArg) {
        return false;
    }

    return constructorArg.__type === TYPE_CONTRACT_REFERENCE
}

function isInstance(constructorArg) {
    return typeof constructorArg === 'object' && constructorArg !== null;
}

```

Reply:
👍